lib/upipe-modules/upipe_vanc_decoder.c

```diff
@@ -0,0 +1,325 @@
+ */
+static void upipe_vanc_decoder_free(struct upipe *upipe)
+{
+    upipe_throw_dead(upipe);
+    upipe_vanc_decoder_clean_uref_mgr(upipe);
+    upipe_vanc_decoder_clean_ubuf_mgr(upipe);
+    upipe_vanc_decoder_clean_output(upipe);
+    upipe_vanc_decoder_clean_urefcount(upipe);
+    upipe_vanc_decoder_free_void(upipe);
+}
+
+/** @internal @This is called when there is new data.
+ *
+ * @param upipe description structure of the pipe
+ * @param uref uref carrying the data
+ * @param upump_p reference to the pump that generated the buffer
+ */
+static void upipe_vanc_decoder_input(struct upipe *upipe,
+                                    struct uref *uref,
+                                    struct upump **upump_p)
+{
+    struct upipe_vanc_decoder *vancd = upipe_vanc_decoder_from_upipe(upipe);
+    if (unlikely(!vancd->uref_mgr || !vancd->ubuf_mgr)) {
+        uref_free(uref);
+        return;
+    }
+
+    const uint8_t *r;
+    int end = -1;
+    if (unlikely(!ubase_check(uref_block_read(uref, 0, &end, &r)))) {
+        upipe_err(upipe, "could not read uref");
+        uref_free(uref);
+        return;
+    }
+
+    if (end < 9) {
+        upipe_dbg_va(upipe, "Packet too small (%d)", end);
+        goto ret;
+    }
+
+    while (end >= 9) {
+        if (r[0] & 0xfc) { /* 000000 */
+            goto ret;
+        }
+
+        bool c_not_y = r[0] & 0x2;
+
+        unsigned line = ((r[0] & 1) << 10) | (r[1] << 2) | (r[2] >> 6);
+        unsigned offset = ((r[2] & 0x3f) << 6) | (r[3] >> 2);
+        uint16_t did = ((r[3] & 0x3) << 8) | r[4];
+        uint16_t sdid = (r[5] << 2) | (r[6] >> 6);
+        uint16_t dc = ((r[6] & 0x3f) << 4) | (r[7] >> 4);
+
+        uint8_t cache = r[7] & 0xf;
+        unsigned cached_bits = 4;
+        r += 8;
+        end -= 8;
+
+        size_t bits_left = 10 * ((dc & 0xff) + 1 /* checksum */) - cached_bits;
+        if (((bits_left + 7) / 8) > end) {
+            upipe_dbg_va(upipe, "Invalid DC %u, packet size %d", dc & 0xff, end);
+            goto ret;
+        }
+
+        end -= (bits_left + 7) / 8;
+
+        if (line == 0) {
+            upipe_dbg(upipe, "Invalid line number 0");
+            r += (bits_left + 7) / 8;
+            continue;
+        }
+
+        struct uref *pic = uref_pic_alloc(vancd->uref_mgr, vancd->ubuf_mgr,
+                S291_HEADER_SIZE + (dc & 0xff) + 1, 1);
+        if (unlikely(!pic)) {
+            upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
+            goto ret;
+        }
+
+        uint8_t *vanc_buf;
+        if (unlikely(!ubase_check(uref_pic_plane_write(pic, "x10",
+                            0, 0, -1, -1, &vanc_buf)))) {
+            uref_free(pic);
+        }
+        uint16_t *data = (uint16_t*)vanc_buf;
+
+        data[0] = S291_ADF1;
+        data[1] = S291_ADF2;
+        data[2] = S291_ADF3;
+        data[3] = did;
+        data[4] = sdid;
+        data[5] = dc;
+
+        for (int i = 0; i < (dc & 0xff) + 1; i++) {
+            if (cached_bits == 0) {
```

Comment:
maybe just turn this into ubits_read?

Reply:
It would not work as initial read is not byte aligned

Reply:
You could read the entire bitstream using ubits_read